File: neetcode/problems/leetcode1472/solution.py
class Node:
    def __init__(self, val: str):
        self.value = val
        self.next = None
        self.prev = None


class BrowserHistory:
    def __init__(self, homepage: str):
        self.curr = Node(homepage)

    def visit(self, url: str) -> None:
        new_node = Node(url)
        new_node.prev = self.curr
        self.curr.next = new_node
        self.curr = new_node

    def back(self, steps: int) -> str:
        while steps > 0 and self.curr.prev:
            self.curr = self.curr.prev
            steps -= 1

        return self.curr.value

    def forward(self, steps: int) -> str:
        while steps > 0 and self.curr.next:
            self.curr = self.curr.next
            steps -= 1

        return self.curr.value

File: neetcode/problems/leetcode1472/solution.js
class Node {
	constructor(val) {
		this.value = val;
		this.prev = null;
		this.next = null;
	}
}

class BrowserHistory {
	constructor(homepage) {
		this.curr = new Node(homepage);
	}

	visit(page) {
		this.curr.next = new Node(page);
		this.curr = this.curr.next;
	}

	back(steps) {
		while (steps > 0 && this.curr.prev) {
			this.curr = this.curr.prev;
			steps--;
		}

		return this.curr.value;
	}

	forward(steps) {
		while (steps > 0 && this.curr.next) {
			this.curr = this.curr.next;
			steps--;
		}

		return this.curr.value;
	}
}

class BrowserHistoryArray {
	constructor(homepage) {
		this.history = [homepage];
		this.curr = 0;
	}

	visit(page) {
		this.history = this.history.slice(0, this.curr + 1).concat(page);
		this.curr = this.history.length - 1;
	}

	back(steps) {
		this.curr = Math.max(this.curr - steps, 0);
		return this.history[this.curr];
	}

	forward(steps) {
		this.curr = Math.min(this.curr + steps, this.history.length - 1);
		return this.history[this.curr];
	}
}

class BrowserHistoryArrayImproved {
	constructor(homepage) {
		this.history = [homepage];
		this.curr = 0;
		this.len = 1;
	}

	visit(page) {
		if (this.history.length < this.curr + 2) this.history.push(page);
		else this.history[this.curr + 1] = page;

		this.curr++;
		this.len = this.curr + 1;
	}

	back(steps) {
		this.curr = Math.max(this.curr - steps, 0);
		return this.history[this.curr];
	}

	forward(steps) {
		this.curr = Math.min(this.curr + steps, this.len - 1);
		return this.history[this.curr];
	}
}
